File: 92. Reverse Linked List II/92.cpp
/*
linked list traversal & multiple maps
time: O(n)
space: O(n)
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* curNode = head;
        unordered_map<int, ListNode*> pos_map;
        unordered_map<ListNode*, ListNode*> prev_map;
        int curpos = 1;
        ListNode* prevNode = nullptr;
        
        while(curNode != nullptr){
            pos_map[curpos++] = curNode;
            prev_map[curNode] = prevNode;
            prevNode = curNode;
            curNode = curNode->next;
        }

        if (left == 1 && curpos> 2){
            curNode = pos_map[right];
        } else{
            curNode = head;
        }

        ListNode* newhead = curNode;

        for(int i=1; i < curpos; i++){
            // within reversing range
            if (i >= left && i < right){
                curNode->next = prev_map[curNode];
            } else if (i == left - 1){
                curNode->next = pos_map[right];
            } else{
                if (pos_map.find(i+1) == pos_map.end()){
                    curNode->next = nullptr;
                } else{
                    curNode->next = pos_map[i+1];
                }
            }

            curNode = curNode->next;
        }

        return newhead;
    }
};

/*
pos_map = {int position : ListNOde* correponding node}
prev_map = {ListNode* node: LisNode* previous node in original list}

1->2->3->4->5->6
*/

File: 92. Reverse Linked List II/92.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
"""
multi pointer: linked list in place reversal
time: O(n)
space: O(1)
"""

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        
        reverseHead = None
        reverseTail = None
        reverseLeft = None
        reverseRight = None
        newHead = None

        counter = 1
        curNode = head

        while curNode != None:
            if counter == left - 1:
                reverseHead = curNode
            elif counter == right + 1:
                reverseTail = curNode
            elif counter == left:
                reverseLeft = curNode
            elif counter == right:
                reverseRight = curNode
            
            curNode = curNode.next
            counter += 1

        prevNode = reverseLeft
        curNode = reverseLeft.next
        counter = left

        while counter < right:
            nextNode = curNode.next
            curNode.next = prevNode
            prevNode = curNode
            curNode = nextNode
            counter += 1

        reverseLeft.next = reverseTail
        if reverseHead != None:
            reverseHead.next = reverseRight

        if left > 1:
            return head
        else:
            return reverseRight

        
"""
edge cases:
if left == right -> return Head

first traversal:
reverseHead = node at original left - 1 position
reverseTail = node at original right + 1 position
reverseLeft = node at original left position
reverseRight = node at original right position

second traversal:
prevNode = node at left
curNode = node at left + 1

while prevNode hasn't reached node at right:
    nextNode = curNode.next
    curNode.next = prevNode
    prevNode = curNode
    curNode = nextNode


reverseLeft.next = reverseTail
reverseHead.next = reverseRight

"""
